src/block-cache.ts

```diff
@@ -0,0 +1,183 @@
+    }
+    // check blockTag
+    const blockTag: string = cacheUtils.blockTagForPayload(payload) as string;
+    if (blockTag === 'pending') {
+      return false;
+    }
+    // can be cached
+    return true;
+  }
+
+  canCacheResult(payload: JsonRpcRequest<string[]>, result: Record<string, unknown>): boolean {
+    // never cache empty values (e.g. undefined)
+    if (emptyValues.includes((result as unknown) as string)) {
+      return false;
+    }
+    // check if transactions have block reference before caching
+    if (['eth_getTransactionByHash', 'eth_getTransactionReceipt'].includes(payload.method)) {
+      if (!result || !result.blockHash || result.blockHash === '0x0000000000000000000000000000000000000000000000000000000000000000') {
+        return false;
+      }
+    }
+    // otherwise true
+    return true;
+  }
+
+  // removes all block caches with block number lower than `oldBlockHex`
+  clearBefore(oldBlockHex: string): void {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const self = this;
+    const oldBlockNumber: number = Number.parseInt(oldBlockHex, 16);
+    // clear old caches
+    Object.keys(self.cache)
+      .map(Number)
+      .filter((num) => num < oldBlockNumber)
+      .forEach((num) => delete self.cache[num]);
+  }
+}
+
+function createBlockCacheMiddleware(
+  opts: BlockCacheMiddlewareOptions = {},
+): JsonRpcMiddleware<string[], Record<string, unknown>> {
+  // validate options
+  const { blockTracker } = opts;
+
+  if (!blockTracker) {
+    throw new Error('createBlockCacheMiddleware - No BlockTracker specified');
+  }
+
+  // create caching strategies
+  const blockCache: BlockCacheStrategy = new BlockCacheStrategy();
+  const strategies: Record<string, BlockCacheStrategy> = {
+    perma: blockCache,
+    block: blockCache,
+    fork: blockCache,
+  };
+
+  return createAsyncMiddleware(async (req, res, next) => {
+    // allow cach to be skipped if so specified
+    if (((req as unknown) as Record<string, unknown>).skipCache) {
+      return next();
+    }
+    // check type and matching strategy
+    const type: string = cacheUtils.cacheTypeForPayload(req);
+    const strategy: BlockCacheStrategy = strategies[type];
+    // If there's no strategy in place, pass it down the chain.
+    if (!strategy) {
+      return next();
+    }
+    // If the strategy can't cache this request, ignore it.
+    if (!strategy.canCacheRequest(req)) {
+      return next();
+    }
+
+    // get block reference (number or keyword)
+    let blockTag: string|null = cacheUtils.blockTagForPayload(req);
+    if (!blockTag) {
+      blockTag = 'latest';
+    }
+
+    // get exact block number
+    let requestedBlockNumber: string;
+    if (blockTag === 'earliest') {
+      // this just exists for symmetry with "latest"
+      requestedBlockNumber = '0x00';
+    } else if (blockTag === 'latest') {
+      // fetch latest block number
+      const latestBlockNumber: string = await blockTracker.getLatestBlock();
```

Comment:
nit: We don't need to explicitly specify here; it will be inferred. There are sometimes DevX and potentially safety benefits to "unnecessarily" specifying types (especially with function/method return types), but I don't think this is such an example.